x-pack/plugins/security/public/management/roles/edit_role/privileges/kibana/privilege_form_calculator/privilege_form_calculator.ts

```diff
@@ -0,0 +1,295 @@
+ */
+
+import { Role } from '../../../../../../../common/model';
+import { isGlobalPrivilegeDefinition } from '../../../privilege_utils';
+import { KibanaPrivileges, SubFeaturePrivilegeGroup } from '../../../../model';
+
+/**
+ * Calculator responsible for determining the displayed and effective privilege values for the following interfaces:
+ * - <PrivilegeSpaceForm> and children
+ * - <PrivilegeSpaceTable> and children
+ */
+export class PrivilegeFormCalculator {
+  constructor(private readonly kibanaPrivileges: KibanaPrivileges, private readonly role: Role) {}
+
+  /**
+   * Returns the assigned base privilege.
+   * If more than one base privilege is assigned, the most permissive privilege will be returned.
+   * If no base privileges are assigned, then this will return `undefined`.
+   *
+   * @param privilegeIndex the index of the kibana privileges role component
+   */
+  public getBasePrivilege(privilegeIndex: number) {
+    const entry = this.role.kibana[privilegeIndex];
+
+    const basePrivileges = this.kibanaPrivileges.getBasePrivileges(entry);
+    return basePrivileges.find(bp => entry.base.includes(bp.id));
+  }
+
+  /**
+   * Returns the *displayed* Primary Feature KibanaPrivilege for the indicated feature and privilege index.
+   * If the effective primary feature privilege is a "minimal" version, then this returns the corresponding non-minimal version.
+   *
+   * @example
+   * The following kibana privilege entry will return `read`:
+   * ```ts
+   * const entry = {
+   *    base: [],
+   *    feature: {
+   *       some_feature: ['minimal_read'],
+   *    }
+   * }
+   * ```
+   *
+   * @param featureId the feature id to get the Primary Feature KibanaPrivilege for.
+   * @param privilegeIndex the index of the kibana privileges role component
+   */
+  public getDisplayedPrimaryFeaturePrivilege(featureId: string, privilegeIndex: number) {
+    const feature = this.kibanaPrivileges.getSecuredFeature(featureId);
+
+    const basePrivilege = this.getBasePrivilege(privilegeIndex);
+
+    const selectedFeaturePrivileges = this.getSelectedFeaturePrivileges(featureId, privilegeIndex);
+
+    return feature.getPrimaryFeaturePrivileges().find(fp => {
+      const correspondingPrivilegeId = fp.getCorrespondingPrivilegeId();
+      const hasMinimalPrivileges = feature.subFeatures.length > 0;
+      return (
+        selectedFeaturePrivileges.includes(fp.id) ||
+        (hasMinimalPrivileges && selectedFeaturePrivileges.includes(correspondingPrivilegeId)) ||
+        basePrivilege?.grantsPrivilege(fp)
+      );
+    });
+  }
+
+  /**
+   * Determines if the indicated feature has sub-feature privileges assigned which are not superseded
+   * by any other assigned privileges.
+   *
+   * @param featureId the feature id
+   * @param privilegeIndex the index of the kibana privileges role component
+   */
+  public hasNonSupersededSubFeaturePrivileges(featureId: string, privilegeIndex: number) {
```

Comment:
That's what I intended (there are no granted sub-feature privileges which superseded the original primary feature privilege), but now that you say it out loud, this could be counter-intuitive.

Perhaps it should be `hasCustomizedSubFeaturePrivileges`, instead of `hasNonSupersededSubFeaturePrivileges`?

Reply:
> `hasCustomizedSubFeaturePrivileges`

That's more along the lines of what I was expecting. It seems like we want to indicate to the user that they should expand the row to see the privileges they were actually granted.

Reply:
Sounds great, I'll make this change